drivers/eeprom/eeprom_fm25w256.c

```diff
@@ -0,0 +1,209 @@
+	const struct spi_buf buf[] = {
+		{
+			.buf = opcode,
+			.len = sizeof(opcode),
+		},
+	};
+
+	struct spi_buf_set tx = {
+		.buffers = buf,
+		.count = 1,
+	};
+
+	const struct config *config = dev->config;
+
+	return spi_write_dt(&config->spi_spec, &tx);
+}
+
+static int memory_operation(const struct device *dev, off_t offset, void *buf, size_t buf_len,
+			    bool is_write)
+{
+	int ret;
+
+	uint8_t opcode[] = {
+		is_write ? OPCODE_WRITE : OPCODE_READ,
+	};
+
+	uint8_t mem_addr[] = {
+		(offset >> 8) & 0xff,
+		offset & 0xff,
+	};
+
+	const struct spi_buf spi_buf[] = {
+		{
+			.buf = opcode,
+			.len = sizeof(opcode),
+		},
+		{
+			.buf = mem_addr,
+			.len = sizeof(mem_addr),
+		},
+		{
+			.buf = buf,
+			.len = buf_len,
+		},
+	};
+
+	struct spi_buf_set tx_buf_set = {
+		.buffers = spi_buf,
+		.count = is_write ? ARRAY_SIZE(spi_buf) : ARRAY_SIZE(spi_buf) - 1,
+	};
+
+	struct spi_buf_set rx_buf_set = {
+		.buffers = spi_buf,
+		.count = ARRAY_SIZE(spi_buf),
+	};
+
+	const struct config *config = dev->config;
+
+	if (is_write && config->read_only) {
+		LOG_ERR("Device is configured read only");
+		return -EACCES;
+	}
+
+	if (offset > SIZE) {
+		LOG_ERR("Offset out of range");
+		return -ENOMEM;
+	}
+
+	if (offset + buf_len > SIZE) {
+		LOG_ERR("Memory roll over");
+		return -EACCES;
+	}
+
+	if (is_write) {
+		ret = write_enable(dev);
+		if (ret < 0) {
+			LOG_ERR("Unable to set write enable latch: %d", ret);
+			return ret;
+		}
+	}
+
+	return spi_transceive_dt(&config->spi_spec, &tx_buf_set, is_write ? NULL : &rx_buf_set);
+}
+
+static int read(const struct device *dev, off_t offset, void *buf, size_t buf_len)
+{
+	LOG_DBG("About to read %u bytes from 0x%04x", buf_len, offset);
+	return memory_operation(dev, offset, buf, buf_len, false);
+}
+
+static int write(const struct device *dev, off_t offset, const void *buf, size_t buf_len)
+{
+	LOG_DBG("About to write %u bytes to 0x%04x", buf_len, offset);
+	return memory_operation(dev, offset, (void *)buf, buf_len, true);
+}
+
+static size_t size(const struct device *dev)
+{
+	return SIZE;
+}
+
+static struct eeprom_driver_api api = {
+	.read = read,
+	.write = write,
+	.size = size,
+};
+
+#define _OPERATION(n)                                                                              \
+	COND_CODE_1(DT_INST_PROP(n, spi_mode_3),                                                   \
+		    SPI_WORD_SET(WORD_SIZE) | SPI_TRANSFER_MSB | SPI_MODE_CPOL | SPI_MODE_CPHA,    \
+		    SPI_WORD_SET(WORD_SIZE) | SPI_TRANSFER_MSB)
+
+#define INIT(n)                                                                                    \
+	static const struct config inst_##n##_config = {                                           \
+		.spi_spec = SPI_DT_SPEC_INST_GET(n, _OPERATION(n), 0),                             \
+		.wp_spec = GPIO_DT_SPEC_INST_GET_OR(n, wp_gpios, { 0 }),                           \
+		.hold_spec = GPIO_DT_SPEC_INST_GET_OR(n, hold_gpios, { 0 }),                       \
```

Comment:
These will set `.wp_spec` and `.hold_spec` to NULL if they fail to find param; shouldn't this rather fail the compilation as it seems that both the `.wp_spec` and `.hold_spec` seem to be required?